lib/chef/client.rb

```diff
@@ -665,6 +683,83 @@ def register(client_name = node_name, config = Chef::Config)
       raise
     end
 
+    #
+    # In the brave new world of No Certs On Disk, we want to put the pem file into Keychain or the Certstore
+    # But is it already there?
+    def check_certstore_for_key(cert_name)
+      require "win32-certstore"
+      win32certstore = ::Win32::Certstore.open("MY")
+      win32certstore.search("#{cert_name}")
+    end
+
+    def generate_pfx_package(cert_name, date = nil)
+      require_relative "mixin/powershell_exec"
+      extend Chef::Mixin::PowershellExec
+      ::Chef::HTTP::Authenticator.get_cert_password
+      powershell_code = <<~EOH
+
+        $date = "#{date}"
+
+        $certSplat = @{
+            Subject = "#{cert_name}"
+            KeyExportPolicy = 'Exportable'
+            KeyUsage = @('KeyEncipherment','DigitalSignature')
+            CertStoreLocation = 'Cert:\\LocalMachine\\My'
+            TextExtension = @("2.5.29.37={text}1.3.6.1.5.5.7.3.2,1.3.6.1.5.5.7.3.1")
+        };
+        if ([string]$date -as [DateTime]){
+          $certSplat.add('NotAfter', $date)
+        }
+
+        New-SelfSignedCertificate @certSplat;
+      EOH
+      powershell_exec!(powershell_code)
+    end
+
+    def move_key_and_register(cert_name)
+      require "time" unless defined?(Time)
+      autoload :URI, "uri"
+
+      KeyMigration.instance.key_migrated = true
+
+      node = Chef::Config[:node_name]
+      d = Time.now
+      end_date = Time.new(d.year, d.month + 3, d.day, d.hour, d.min, d.sec).utc.iso8601
+
+      payload = {
+        name: node,
+        clientname: node,
+        public_key: "",
+        expiration_date: end_date,
+      }
+
+      generate_pfx_package(cert_name, end_date)
+      payload[:public_key] = get_public_key(cert_name)
+      base_url = "#{Chef::Config[:chef_server_url]}"
+      client = Chef::ServerAPI.new(base_url, client_name: Chef::Config[:validation_client_name], signing_key_filename: Chef::Config[:validation_key])
+      client.post(base_url + "/clients", payload)
+      KeyMigration.instance.key_migrated = false
```

Comment:
So what happens at the top of that method is that I set the flag Keymigration to True - Line 739 executes and calls code in Authenticator.rb to load the pem file needed to build the headers for the client object. In this case, I need to validator so code in authenticator sees the Keymigraiton flag and hands me back the validator pem. On line 740 I update the chef server with the new key for the client. I now need the chef client to connect with the new cert I stuffed into the Certstore (line 738 generate_pfx_package) so I set the Keymigraiton flag back to false to prevent the Load_Signing_key code in authenticator from handing me back the validator pem when the client goes on to run a CCR

Reply:
This makes a lot more sense.